lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,662 @@
+            PreloadHeroImageConfiguration::INLINE_STYLE_BACKUP_ATTRIBUTE
+        );
+
+        $heroImages     = $this->findHeroImages($document);
+        $heroImageCount = count($heroImages);
+        if ($heroImageCount > self::DATA_HERO_MAX) {
+            $errors->add(Error\TooManyHeroImages::whenPastMaximum());
+            $heroImageCount = self::DATA_HERO_MAX;
+        }
+
+        for ($index = 0; $index < $heroImageCount; $index++) {
+            $this->generatePreload($heroImages[$index], $document, $errors);
+            $this->generateImg($heroImages[$index], $document);
+        }
+    }
+
+    /**
+     * Find the hero images to optimize.
+     *
+     * @param Document $document Document to look for hero images in.
+     * @return HeroImage[] Array of hero images to optimize.
+     */
+    private function findHeroImages(Document $document)
+    {
+        $heroImageCandidate = null;
+        $heroImages         = [];
+        $node               = $document->body;
+
+        while ($node !== null) {
+            if (! $node instanceof DOMElement) {
+                $node = $this->nextNode($node);
+                continue;
+            }
+
+            $heroImage = $this->detectImageWithDataHero($node);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            }
+
+            if (! $heroImageCandidate && count($heroImages) === 0) {
+                $heroImageCandidate = $this->detectHeroImageCandidate($node);
+            }
+            if (Amp::isTemplate($node)) {
+                // Ignore images inside templates.
+                $node = $this->skipNodeAndChildren($node);
+            } else {
+                $node = $this->nextNode($node);
+            }
+        }
+
+        // Optimize data-hero element if defined.
+        if (count($heroImages) > 0) {
+            return $heroImages;
+        }
+
+        // Fall back to auto-detected hero image if available.
+        if ($heroImageCandidate) {
+            return [$heroImageCandidate];
+        }
+
+        // No hero images to optimize.
+        return [];
+    }
+
+    /**
+     * Detect a hero image with the data-hero attribute.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image, or null if none detected.
+     */
+    private function detectImageWithDataHero(DOMElement $element)
+    {
+        if (!$element->hasAttribute(Attribute::DATA_HERO)) {
+            return null;
+        }
+
+        if ($element->tagName === Extension::IMAGE) {
+            return new HeroImage(
+                $element->getAttribute(Attribute::SRC),
+                $element->getAttribute(Attribute::MEDIA),
+                $element->getAttribute(Attribute::SRCSET),
+                $element
+            );
+        }
+
+        if ($this->isAmpEmbed($element)) {
+            return $this->getPlaceholderImage($element);
```

Comment:
The element itself may have a background instead of having a placeholder, so this will account for that:

```suggestion
            $placeholderImage = $this->getPlaceholderImage($element);
            if (null !== $placeholderImage) {
                return $placeholderImage;
            }
```